partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt

```diff
@@ -760,6 +762,54 @@ internal class PlanTyper(
             return rex(type, op)
         }
 
+        // COALESCE(v1, v2,..., vN)
+        // ==
+        // CASE
+        //     WHEN v1 IS NOT NULL THEN v1  -- WHEN branch always a boolean
+        //     WHEN v2 IS NOT NULL THEN v2  -- WHEN branch always a boolean
+        //     ... -- similarly for v3..vN-1
+        //     ELSE vN
+        // END
+        // --> minimal common supertype of(<type v1>, <type v2>, ..., <type v3>)
+        override fun visitRexOpCoalesce(node: Rex.Op.Coalesce, ctx: StaticType?): Rex {
+            val args = node.args.map { visitRex(it, it.type) }.toMutableList()
+            val typer = DynamicTyper()
+            args.forEach { v ->
+                typer.accumulate(v.type)
+            }
+            val (type, mapping) = typer.mapping()
+            if (mapping != null) {
+                assert(mapping.size == args.size) { "Coercion mappings `len ${mapping.size}` did not match the number of COALESCE arguments `len ${args.size}`" }
+                for (i in args.indices) {
+                    val (operand, target) = mapping[i]
+                    if (operand == target) continue // skip; no coercion needed
+                    val cast = env.fnResolver.cast(operand, target)
+                    val rex = rex(type, rexOpCallStatic(fnResolved(cast), listOf(args[i])))
+                    args[i] = rex
+                }
+            }
+            val op = rexOpCoalesce(args)
+            return rex(type, op)
+        }
+
+        // NULLIF(v1, v2)
+        // ==
+        // CASE
+        //     WHEN v1 = v2 THEN NULL -- WHEN branch always a boolean
+        //     ELSE v1
+        // END
+        // --> minimal common supertype of (NULL, <type v1>)
+        override fun visitRexOpNullif(node: Rex.Op.Nullif, ctx: StaticType?): Rex {
+            val v1 = visitRex(node.v1, node.v1.type)
+            val v2 = visitRex(node.v2, node.v2.type)
+            val typer = DynamicTyper()
+            typer.accumulate(NULL)
+            typer.accumulate(v1.type)
```

Comment:
Only add `v1`'s type here. I found that posgresql actually does something different here. Per [their docs](https://www.postgresql.org/docs/current/functions-conditional.html#FUNCTIONS-NULLIF):
> The result has the same type as the first argument — but there is a subtlety. What is actually returned is the first argument of the implied = operator, and in some cases that will have been promoted to match the second argument's type. For example, NULLIF(1, 2.2) yields numeric, because there is no integer = numeric operator, only numeric = numeric.

Per my read of [SQL-99](https://web.cecs.pdx.edu/~len/sql1999.pdf#page=205), I can't find any mention of this promotion behavior, so leaving the result type as just the minimal common supertype of (\<type v1\>, null). Postgresql would do something like the minimal common supertype of (\<type v1\>, \<type v2\>, null).

Reply:
Yeah I'll simplify it. I initially included the dynamic typer in case we wanted to follow what postgresql did, which is easy to do w/ the dynamic typer.

Reply:
Followup from offline discussion, will keep the dynamic typer here to ensure consistency with the typing of `CASE WHEN` and `COALESCE`.